### ISSUE TITLE
Project Instruction 

### ISSUE THREAD

**Issue by jentle** [saizk/Deep-Learning-for-Solar-Panel-Recognition#1]
Hi,

Great work after reading the presentation pdf inside reports directory. But the code seems not work and lack of instructions about how to access the data and train the models. Commands in Makefile are corrupted like "make data" or it is synced from s3? How can I run this project. I also noticed that the dataset is consists of different resolution pv images. How to handle the resolution difference in training?

Thank so much!

**Comment by saizk** [comment]
Hi! I am glad that the project is of interest to you.
Running the code might result a bit messy, since we had a tight deadline when we developed it. I'll try my best to explain it since it has been a long time and I have removed most of the data from my computer.

**DATA SOURCES**
The **Makefile** was generated by the cookiecutter and has not been used in the project. It is already deleted from the repo.

The **main data source** can be found in the README ([https://zenodo.org/record/5171712](url)) and contains more than 3,500 images. The resolution of the images is 256x256 pixels but have different **spatial** resolutions. In other words, the images have the same size but are taken from different altitudes. We did not tackle this problem directly, as most of the networks architectures have internal features to do it, i.e. FPN (Feature Pyramid Network).

Regarding **Google Maps images**, the resolution at which they were scraped was 1280x1280 but later were divided in 25 different images of 256x256. You can find the respective code in _src/data/wrappers.py_ inside the `GoogleMapsAPIDownloader` object. The method is called `split_tile()` and it is done automatically when you call `download_tile()` and set `split=True`.

**OBJECT DETECTION**
Object Detection was done using **YOLOv5** [https://github.com/ultralytics/yolov5](url). The data preprocessing can be found in _src/features/yolo_preprocess_data.py_ or _src/models/yolo/yolov5/data/preprocess_data.py_ but I think that both are outdated versions. In the end what we did was converting all the images to PNG and splitting them in two different folders: **images**, which contains the raw aerial images and **masks**, which contains the respective ground truth images. Each of these folders contain three different subfolders: **train**, **test** and **val**.

An additional folder called **labels** was created with the same internal structure. The annotations were computed using _create_yolo_annotations.py_ (found also in _src/features_ and _src/models/yolo/yolov5/data_).

```
├── solar_panels
     ├── images
     │     ├── train
     │     ├── test
     │     └── val
     ├── masks
     │     ├── train
     │     ├── test
     │     └── val
     └── labels  # generated with create_yolo_annotations.py
           ├── train
           ├── test
           └── val
```
You can specify custom paths with your own structure inside **_sp_dataset.yaml_** (_src/models/yolo_).

**SEGMENTATION**
The segmentation part was developed using **segmentation_models.pytorch** repository [https://github.com/qubvel/segmentation_models.pytorch](url). You can find an easy way to run the code inside the _notebooks_ folder. Two versions were written to train the models. The first one is a vanilla implementation using PyTorch (_segmentation_pytorch.ipynb_) and the second one is equivalent to the first one but using PyTorch Lightning to facilitate readability and future improvements (_segmentation_pytorch_lightning.ipynb_).

I do not exactly remember how the data was structured but as far as I can read in _src/models/segmentation/datasets.py_, the images and their labels were separated in different folders. I guess that it is structured in the same way as in the object detection part. The naming convention was maintained from the original source (mask images ended with "__label_").

**MODELS**
I have updated the repo with some light models both from object detection and segmentation. If you are interested in the rest of them I can upload them to Google Drive.

I hope that this information helps you. In case you have any further doubts, feel free to ask them in this issue.

**Comment by jentle** [comment]
Hi, Sergio

Thank you so much for the detailed explanation for the project and quick update. I’m working on similar area and your project inspired me a lot.  I’ll try to build it following the instructions. From the presentation pdf, the project has compared result for different 
object detection and segmentation architecture which I can train with my own. Thanks a lot if you can provide the rest model info if available. So that l can take it as baseline and saving a lot of time.

**Comment by saizk** [comment]
I have updated the README file with some basic instructions to run the code and I have added two links in the **MODELS** section to download the respective models.

Hope it works!

**Comment by jentle** [comment]
Thanks!

I'm working on data processing. I guess some key actions was [commented](https://github.com/saizk/Deep-Learning-for-Solar-Panel-Recognition/blob/83570c659592054fcfd858bc3fb2e8a433d5d786/src/features/yolo_preprocess_data.py#L68).  It's ok for me right now. The notebook might be a better option for exploration.

**Comment by saizk** [comment]
Yes, there might be some useful code commented, sorry for that.

I mark this issue as completed.
